rclrs/src/node.rs

```diff
@@ -71,10 +71,15 @@ pub struct Node {
 /// This struct manages the lifetime of an `rcl_node_t`, and accounts for its
 /// dependency on the lifetime of its `rcl_context_t` by ensuring that this
 /// dependency is [dropped after][1] the `rcl_node_t`.
+/// Note: we capture the rcl_node_t returned from rcl_get_zero_initialized_node()
+/// to guarantee that the node handle exists until we drop the NodeHandle
+/// instance. This addresses an issue where previously the address of the variable
+/// in the builder.rs was being used, and whose lifespan was (just) shorter than the
+/// NodeHandle instance.
 ///
 /// [1]: <https://doc.rust-lang.org/reference/destructors.html>
 pub(crate) struct NodeHandle {
-    pub(crate) rcl_node: Mutex<rcl_node_t>,
+    pub(crate) rcl_node: Mutex<Box<rcl_node_t>>,
```

Comment:
Can I check why this was put into a Box? The contents of `rcl_node` are just pointers so there shouldn't be any issues with moving the address of the `rcl_node` (and its parent struct is managed by an Arc, so its address shouldn't change anyway).

Reply:
Good question - without this change it appears as though the node handle goes out of scope before we get chance to clear the logging publisher (in humble this occurs in the rcl_node_fini() call).  The behaviour without the change is that you will get a segfault on node closure because the object pointed to by the rcl_node_t is no longer valid.  Took a bit of debugging to get to the bottom of it - and honestly, I don't particularly like the change with the extra de-refs.
Looking at Jazzy, I believe that the issue will no longer occur due the logging publisher being managed outside the rcl_node_init/fini calls.
For reference the failure occurs [here](https://github.com/ros2/rcl/blob/6d52f2b54980140476222b420467e7fbd8575dd9/rcl/src/rcl/node.c#L378)

Reply:
Thanks for the explanation. I think we do want to keep supporting humble for now, so I'll try to reproduce the seg fault in humble and see if there may be alternative solution.

I'm still pretty lost on how boxing the `rcl_node_t` would affect its lifetime, but if I can reproduce the problem then I should be able to make sense of it.

Reply:
Oh actually if rcl itself is storing the `rcl_node_t*` value somewhere and using it without consideration for its lifetime then I can see how this would happen. I'll check if that's the case, and maybe we can set things up so that we don't initialize the `rcl_node_t` until it's already inside the `NodeHandle`.

Reply:
Yep, that's it in a nutshell - the rosout code keeps a record of the node used to create it (this [line](https://github.com/ros2/rcl/blob/6d52f2b54980140476222b420467e7fbd8575dd9/rcl/src/rcl/logging_rosout.c#L189) specifically) and this pointed to the stack variable in NodeHandle ([here](https://github.com/ros2-rust/ros2_rust/blob/f45a66f47dc727e3ccb13037a6c57923af1446c7/rclrs/src/node/builder.rs#L280)).  During the `rcl_node_fini` we call `rcl_logging_rosout_fini_publisher_for_node` which tries to reference the stack pointer, however, this area of memory it points to has now been mangled, i.e. the pointer still points to the same memory location but the values at that spot are garbage.

Reply:
I've gone ahead and put together an alternative fix here: https://github.com/geoff-imdex/ros2_rust/pull/5